nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/CSVValidators.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.nifi.components.ValidationContext;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.components.Validator;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class CSVValidators {
+
+    public static class SingleCharacterValidator implements Validator {
+        private static final Set<String> illegalChars = new HashSet<>();
+
+        static {
+            illegalChars.add("\r");
+            illegalChars.add("\n");
+        }
+
+        @Override
+        public ValidationResult validate(final String subject, final String input, final ValidationContext context) {
+
+            if (input == null) {
+                return new ValidationResult.Builder()
+                        .input(input)
+                        .subject(subject)
+                        .valid(false)
+                        .explanation("Input is null for this property")
+                        .build();
+            }
+
+            final String unescaped = CSVUtils.unescape(input);
+            if (unescaped.length() != 1) {
+                return new ValidationResult.Builder()
+                        .input(input)
+                        .subject(subject)
+                        .valid(false)
+                        .explanation("Value must be exactly 1 character but was " + input.length() + " in length")
+                        .build();
+            }
+
+            if (illegalChars.contains(unescaped)) {
+                return new ValidationResult.Builder()
+                        .input(input)
+                        .subject(subject)
+                        .valid(false)
+                        .explanation(input + " is not a valid character for this property")
+                        .build();
+            }
+
+            return new ValidationResult.Builder()
+                    .input(input)
+                    .subject(subject)
+                    .valid(true)
+                    .build();
+        }
+
+    }
+
+    public static final Validator UNESCAPED_SINGLE_CHAR_VALIDATOR = new Validator() {
+        @Override
+        public ValidationResult validate(final String subject, final String input, final ValidationContext context) {
+
+            if (input == null) {
+                return new ValidationResult.Builder()
+                        .input(input)
+                        .subject(subject)
+                        .valid(false)
+                        .explanation("Input is null for this property")
+                        .build();
+            }
+
+            String unescapeString = unescapeString(input);
+
+            return new ValidationResult.Builder()
+                    .subject(subject)
+                    .input(unescapeString)
+                    .explanation("Only non-null single characters are supported")
+                    .valid((input.length() == 1 && input.charAt(0) != 0) || context.isExpressionLanguagePresent(input))
```

Comment:
Shouldn't this be checking unescapeString instead of input? If I put "\u0001" as the value of the character, unescapeString() gets the correct value, but then the property is marked invalid because the original input was 6 characters long. The SingleCharacterValidator checks the "unescaped" variable instead of the original input, for this reason.

Reply:
@mattyb149 Terribly sorry about this. This is super embarrassing. Was wondering how the testcases failed me. It should have been "\\u0001" instead of "\u0001".  Fixed the testcase and the class. Will run a manual test on the UI as well before pushing the changes.

Reply:
I hadn't touched the unit test, and I am having trouble pushing my branch, so please feel free to update and run your tests, and I will use this PR as-is.

Reply:
Ironically I had been trying to emulate your unit test by pasting in the actual character for \u0001, but that doesn't make it through, an empty string gets passed to the validator. That's when I tried the actual string "\u0001" and how I found the issue.